File: tests/lua/xpb_test.lua
-- XPB V2 Lua Runtime Tests
-- Tests round-trip encoding/decoding for all types

local xpb = require "xpb"

local tests_passed = 0
local tests_failed = 0

local function test(name, cond)
    if cond then
        print(string.format("  [PASS] %s", name))
        tests_passed = tests_passed + 1
    else
        print(string.format("  [FAIL] %s", name))
        tests_failed = tests_failed + 1
    end
end

-- Test bool round-trip
local function test_bool()
    print("\n=== Test Bool ===")

    local enc = xpb.Encoder(64)
    enc:write_bool(true)
    enc:write_bool(false)
    local data = enc:finish()

    local dec = xpb.Decoder(data)
    test("bool true", dec:read_bool() == true)
    test("bool false", dec:read_bool() == false)
end

-- Test int32 round-trip
local function test_int32()
    print("\n=== Test Int32 ===")

    local values = {0, 1, -1, 100, -100, 2147483647, -2147483648}
    local enc = xpb.Encoder(256)
    for i, v in ipairs(values) do
        enc:write_int32(v)
    end
    local data = enc:finish()

    local dec = xpb.Decoder(data)
    for i, v in ipairs(values) do
        test(string.format("int32 value %d", v), dec:read_int32() == v)
    end
end

-- Test int64 round-trip
local function test_int64()
    print("\n=== Test Int64 ===")

    local values = {0, 1, -1, 1000000000, -1000000000, 9223372036854775807, -9223372036854775807}
    local enc = xpb.Encoder(256)
    for i, v in ipairs(values) do
        enc:write_int64(v)
    end
    local data = enc:finish()

    local dec = xpb.Decoder(data)
    for i, v in ipairs(values) do
        test(string.format("int64 value %d", v), dec:read_int64() == v)
    end
end

-- Test float32 round-trip
local function test_float32()
    print("\n=== Test Float32 ===")

    local values = {0.0, 1.0, -1.0, 3.14159, -273.15}
    local enc = xpb.Encoder(256)
    for i, v in ipairs(values) do
        enc:write_float32(v)
    end
    local data = enc:finish()

    local dec = xpb.Decoder(data)
    for i, v in ipairs(values) do
        local decoded = dec:read_float32()
        test(string.format("float32 value %f", v), math.abs(decoded - v) < 0.0001)
    end
end

-- Test float64 round-trip
local function test_float64()
    print("\n=== Test Float64 ===")

    local values = {0.0, 1.0, -1.0, 3.14159265358979, -273.15, 1e100}
    local enc = xpb.Encoder(256)
    for i, v in ipairs(values) do
        enc:write_float64(v)
    end
    local data = enc:finish()

    local dec = xpb.Decoder(data)
    for i, v in ipairs(values) do
        local decoded = dec:read_float64()
        test(string.format("float64 value %g", v), math.abs(decoded - v) < 1e-10)
    end
end

-- Test string round-trip
local function test_string()
    print("\n=== Test String ===")

    local values = {"", "a", "hello", "hello world", "1234567890", "This is a longer string with many characters"}
    for i, v in ipairs(values) do
        local enc = xpb.Encoder(256)
        enc:write_string(v)
        local data = enc:finish()

        local dec = xpb.Decoder(data)
        local decoded = dec:read_string()
        test(string.format("string '%s'", v), decoded == v)
    end

    -- Test long string (>254 chars)
    local long_str = string.rep("x", 300)
    local enc = xpb.Encoder(512)
    enc:write_string(long_str)
    local data = enc:finish()

    local dec = xpb.Decoder(data)
    local decoded = dec:read_string()
    test("long string (>254 chars) length", #decoded == 300)
    test("long string content", decoded == long_str)
end

-- Test bytes round-trip
local function test_bytes()
    print("\n=== Test Bytes ===")

    local data1 = string.char(0x01, 0x02, 0x03, 0x04, 0x05)

    -- Generate 256 bytes
    local data2 = ""
    for i = 0, 255 do
        data2 = data2 .. string.char(i)
    end

    local enc = xpb.Encoder(512)
    enc:write_bytes(data1)
    enc:write_bytes(data2)
    local data = enc:finish()

    local dec = xpb.Decoder(data)
    local decoded1 = dec:read_bytes()
    local decoded2 = dec:read_bytes()

    test("small bytes length", #decoded1 == #data1)
    test("small bytes content", decoded1 == data1)
    test("large bytes length", #decoded2 == #data2)
    test("large bytes content", decoded2 == data2)
end

-- Test nested messages
local function test_nested_message()
    print("\n=== Test Nested Message ===")

    -- Encode inner message
    local inner_enc = xpb.Encoder(64)
    inner_enc:write_string("inner_value")
    inner_enc:write_int32(42)
    local inner_data = inner_enc:finish()

    -- Encode outer message
    local outer_enc = xpb.Encoder(256)
    outer_enc:write_string("outer_name")
    outer_enc:write_message(inner_data)
    local outer_data = outer_enc:finish()

    -- Decode
    local dec = xpb.Decoder(outer_data)
    local name = dec:read_string()
    local inner_out = dec:read_message_bytes()

    test("outer string", name == "outer_name")
    test("inner message length", #inner_out == #inner_data)

    local inner_dec = xpb.Decoder(inner_out)
    local inner_str = inner_dec:read_string()
    local inner_int = inner_dec:read_int32()

    test("inner string", inner_str == "inner_value")
    test("inner int", inner_int == 42)
end

-- Test all types combined
local function test_all_types()
    print("\n=== Test All Types Combined ===")

    local enc = xpb.Encoder(256)
    enc:write_bool(true)
    enc:write_int32(-12345)
    enc:write_int64(9876543210)
    enc:write_float32(3.14)
    enc:write_float64(2.718281828)
    enc:write_string("test string")
    enc:write_bytes(string.char(0xDE, 0xAD, 0xBE, 0xEF))
    local data = enc:finish()

    local dec = xpb.Decoder(data)
    test("bool", dec:read_bool() == true)
    test("int32", dec:read_int32() == -12345)
    test("int64", dec:read_int64() == 9876543210)
    test("float32", math.abs(dec:read_float32() - 3.14) < 0.001)
    test("float64", math.abs(dec:read_float64() - 2.718281828) < 1e-9)
    test("string", dec:read_string() == "test string")
    local bytes = dec:read_bytes()
    test("bytes length", #bytes == 4)
    test("bytes content", bytes == string.char(0xDE, 0xAD, 0xBE, 0xEF))
end

-- Run all tests
print("===========================================")
print("XPB V2 Lua Runtime Tests")
print("===========================================")

test_bool()
test_int32()
test_int64()
test_float32()
test_float64()
test_string()
test_bytes()
test_nested_message()
test_all_types()

print("\n===========================================")
print(string.format("Results: %d passed, %d failed", tests_passed, tests_failed))
print("===========================================")

if tests_failed > 0 then
    os.exit(1)
end

File: tests/java/XpbTest.java
package xpb;

public class XpbTest {
    private static int testsPassed = 0;
    private static int testsFailed = 0;

    private static void test(String name, boolean cond) {
        if (cond) {
            System.out.println("  [PASS] " + name);
            testsPassed++;
        } else {
            System.out.println("  [FAIL] " + name);
            testsFailed++;
        }
    }

    private static void testBool() {
        System.out.println("\n=== Test Bool ===");
        Encoder enc = new Encoder(64);
        enc.writeBool(true);
        enc.writeBool(false);
        byte[] data = enc.finish();

        Decoder dec = new Decoder(data);
        test("bool true", dec.readBool() == true);
        test("bool false", dec.readBool() == false);
    }

    private static void testInt32() {
        System.out.println("\n=== Test Int32 ===");
        int[] values = {0, 1, -1, 100, -100, 2147483647, -2147483648};
        Encoder enc = new Encoder(256);
        for (int v : values) {
            enc.writeInt32(v);
        }
        byte[] data = enc.finish();

        Decoder dec = new Decoder(data);
        for (int i = 0; i < values.length; i++) {
            test("int32 value " + values[i], dec.readInt32() == values[i]);
        }
    }

    private static void testInt64() {
        System.out.println("\n=== Test Int64 ===");
        long[] values = {0L, 1L, -1L, 1000000000L, -1000000000L, 9223372036854775807L, -9223372036854775807L};
        Encoder enc = new Encoder(256);
        for (long v : values) {
            enc.writeInt64(v);
        }
        byte[] data = enc.finish();

        Decoder dec = new Decoder(data);
        for (int i = 0; i < values.length; i++) {
            test("int64 value " + values[i], dec.readInt64() == values[i]);
        }
    }

    private static void testFloat32() {
        System.out.println("\n=== Test Float32 ===");
        float[] values = {0.0f, 1.0f, -1.0f, 3.14159f, -273.15f};
        Encoder enc = new Encoder(256);
        for (float v : values) {
            enc.writeFloat32(v);
        }
        byte[] data = enc.finish();

        Decoder dec = new Decoder(data);
        for (int i = 0; i < values.length; i++) {
            float decoded = dec.readFloat32();
            test("float32 value " + values[i], Math.abs(decoded - values[i]) < 0.0001f);
        }
    }

    private static void testFloat64() {
        System.out.println("\n=== Test Float64 ===");
        double[] values = {0.0, 1.0, -1.0, 3.14159265358979, -273.15, 1e100};
        Encoder enc = new Encoder(256);
        for (double v : values) {
            enc.writeFloat64(v);
        }
        byte[] data = enc.finish();

        Decoder dec = new Decoder(data);
        for (int i = 0; i < values.length; i++) {
            double decoded = dec.readFloat64();
            test("float64 value " + values[i], Math.abs(decoded - values[i]) < 1e-10);
        }
    }

    private static void testString() {
        System.out.println("\n=== Test String ===");
        String[] values = {"", "a", "hello", "hello world", "1234567890", "This is a longer string with many characters"};
        for (String v : values) {
            Encoder enc = new Encoder(256);
            enc.writeString(v);
            byte[] data = enc.finish();

            Decoder dec = new Decoder(data);
            String decoded = dec.readString();
            test("string '" + v + "'", decoded.equals(v));
        }

        // Test long string (>254 chars)
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < 300; i++) {
            sb.append('x');
        }
        String longStr = sb.toString();

        Encoder enc = new Encoder(512);
        enc.writeString(longStr);
        byte[] data = enc.finish();

        Decoder dec = new Decoder(data);
        String decoded = dec.readString();
        test("long string (>254 chars) length", decoded.length() == 300);
        test("long string content", decoded.equals(longStr));
    }

    private static void testBytes() {
        System.out.println("\n=== Test Bytes ===");
        byte[] data1 = new byte[] {0x01, 0x02, 0x03, 0x04, 0x05};
        byte[] data2 = new byte[256];
        for (int i = 0; i < 256; i++) {
            data2[i] = (byte)i;
        }

        Encoder enc = new Encoder(512);
        enc.writeBytes(data1);
        enc.writeBytes(data2);
        byte[] data = enc.finish();

        Decoder dec = new Decoder(data);
        byte[] decoded1 = dec.readBytes();
        byte[] decoded2 = dec.readBytes();

        test("small bytes length", decoded1.length == data1.length);
        test("small bytes content", java.util.Arrays.equals(decoded1, data1));
        test("large bytes length", decoded2.length == data2.length);
        test("large bytes content", java.util.Arrays.equals(decoded2, data2));
    }

    private static void testNestedMessage() {
        System.out.println("\n=== Test Nested Message ===");

        // Encode inner message
        Encoder innerEnc = new Encoder(64);
        innerEnc.writeString("inner_value");
        innerEnc.writeInt32(42);
        byte[] innerData = innerEnc.finish();

        // Encode outer message
        Encoder outerEnc = new Encoder(256);
        outerEnc.writeString("outer_name");
        outerEnc.writeMessage(innerData);
        byte[] outerData = outerEnc.finish();

        // Decode
        Decoder dec = new Decoder(outerData);
        String name = dec.readString();
        byte[] innerOut = dec.readMessageBytes();

        test("outer string", name.equals("outer_name"));
        test("inner message length", innerOut.length == innerData.length);

        Decoder innerDec = new Decoder(innerOut);
        String innerStr = innerDec.readString();
        int innerInt = innerDec.readInt32();

        test("inner string", innerStr.equals("inner_value"));
        test("inner int", innerInt == 42);
    }

    private static void testAllTypes() {
        System.out.println("\n=== Test All Types Combined ===");
        Encoder enc = new Encoder(256);
        enc.writeBool(true);
        enc.writeInt32(-12345);
        enc.writeInt64(9876543210L);
        enc.writeFloat32(3.14f);
        enc.writeFloat64(2.718281828);
        enc.writeString("test string");
        enc.writeBytes(new byte[] {(byte)0xDE, (byte)0xAD, (byte)0xBE, (byte)0xEF});
        byte[] data = enc.finish();

        Decoder dec = new Decoder(data);
        test("bool", dec.readBool() == true);
        test("int32", dec.readInt32() == -12345);
        test("int64", dec.readInt64() == 9876543210L);
        test("float32", Math.abs(dec.readFloat32() - 3.14f) < 0.001f);
        test("float64", Math.abs(dec.readFloat64() - 2.718281828) < 1e-9);
        test("string", dec.readString().equals("test string"));
        byte[] bytes = dec.readBytes();
        test("bytes length", bytes.length == 4);
        test("bytes content", java.util.Arrays.equals(bytes, new byte[] {(byte)0xDE, (byte)0xAD, (byte)0xBE, (byte)0xEF}));
    }

    public static void main(String[] args) {
        System.out.println("===========================================");
        System.out.println("XPB V2 Java Runtime Tests");
        System.out.println("===========================================");

        testBool();
        testInt32();
        testInt64();
        testFloat32();
        testFloat64();
        testString();
        testBytes();
        testNestedMessage();
        testAllTypes();

        System.out.println("\n===========================================");
        System.out.println("Results: " + testsPassed + " passed, " + testsFailed + " failed");
        System.out.println("===========================================");

        if (testsFailed > 0) {
            System.exit(1);
        }
    }
}
